### PR TITLE
DDP communication hook: skip dividing grads by world_size if hook registered.

### PR DESCRIPTION
**Pull request by sinannasir**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42400 DDP communication hook: skip dividing grads by world_size if hook registered.**

@mcarilli spotted that in the original DDP communication hook design described in [39272](https://github.com/pytorch/pytorch/issues/39272), the hooks receive grads that are already predivided by world size.

It makes sense to skip the divide completely if hook registered. The hook is meant for the user to completely override DDP communication. For example, if the user would like to implement something like GossipGrad, always dividing by the world_size would not be a good idea.

We also included a warning in the register_comm_hook API as:
> GradBucket bucket's tensors will not be predivided by world_size. User is responsible to divide by the world_size in case of operations like allreduce.

**Update:** We discovered and fixed a bug with the sparse tensors case. See new unit test called `test_ddp_comm_hook_sparse_gradients` and changes in `reducer.cpp`.

Differential Revision: [D22883905](https://our.internmc.facebook.com/intern/diff/D22883905/)

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -639,6 +639,10 @@ def _register_comm_hook(self, state: object, hook: callable):
                              torch._C.Future associated with the completion of
                              c10d.ProcessGroupNCCL.work.
 
+        .. warning ::
+            GradBucket bucket's tensors will not be predivided by world_size. User is
```
Comment (pritamdamania87) [inline torch/nn/parallel/distributed.py:643]: nit: mentioning bucket twice here, probably say GradBucket's tensors will ....

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 2]
```diff
@@ -385,7 +391,10 @@ void Reducer::mark_variable_ready_sparse(VariableIndex index) {
     // struct are empty, and there is no pre-existing accumulation tensor.
     // Directly assign the sparse tensor to the `contents` field.
     replica.contents = grad;
-    replica.contents.div_(process_group_->getSize());
+    // See Note [DDP Communication Hook]
+    if (comm_hook_ == nullptr) {
+      replica.contents.div_(process_group_->getSize());
+    }
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/reducer.cpp:397]: We should probably test the communication hook with sparse tensors as well. Using [nn.EmbeddingBag](https://pytorch.org/docs/stable/generated/torch.nn.EmbeddingBag.html) with `sparse=True` will generate sparse gradients for you.

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
Looks good overall, requesting changes since I just realized that we don't test the communication hook with sparse tensors.

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (sinannasir) [reply torch/nn/parallel/distributed.py:643]: just replaced with grad bucket to be consistent with other parts of the API.

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 2 above)
Comment (sinannasir) [reply torch/csrc/distributed/c10d/reducer.cpp:397]: @pritamdamania87 Thanks for this comment! It was quite useful to discover and fix a bug with sparse tensors case. 
I think we should just copy in case of sparse gradients, since `bucket_view` is not used. Please see `test_ddp_comm_hook_sparse_gradients` and changes in `reducer.cpp` to make it work.

**Review by pritamdamania87 — APPROVED** [review]
